network/data_link/at_parser.c

```diff
@@ -0,0 +1,1145 @@
+					(int32_t)param->connection.local_port,
+					(int32_t)param->connection.udp_mode);
+			}
+		}
+		break;
+	case AT_SEND:
+		conn_id = desc->multiple_conections ? param->connection.id : 0;
+		if (desc->conn[conn_id].type == SOCKET_TCP) {
+			if (desc->multiple_conections)
+				at_sprintf(&desc->cmd, PUI8("dd"),
+					   (int32_t)param->send_data.id,
+					   (int32_t)param->send_data.data.len);
+			else
+				at_sprintf(&desc->cmd, PUI8("d"),
+					   (int32_t)param->send_data.data.len);
+		} else {
+			if (desc->multiple_conections)
+				at_sprintf(
+					&desc->cmd, PUI8("ddsd"),
+					(int32_t)param->send_data.id,
+					(int32_t)param->send_data.data.len,
+					&param->send_data.remote_ip,
+					(int32_t)param->send_data.remote_port);
+			else
+				at_sprintf(
+					&desc->cmd, PUI8("dsd"),
+					(int32_t)param->send_data.data.len,
+					&param->send_data.remote_ip,
+					(int32_t)param->send_data.remote_port);
+		}
+		break;
+	case AT_STOP_CONNECTION:
+		at_sprintf(&desc->cmd, PUI8("d"), param->conn_id);
+		break;
+	case AT_SET_CONNECTION_TYPE:
+		at_sprintf(&desc->cmd, PUI8("d"),
+			   (int32_t)param->conn_type);
+		break;
+	case AT_SET_SERVER:
+		at_sprintf(&desc->cmd, PUI8("dd"),
+			   (int32_t)param->server.action,
+			   (int32_t)param->server.port);
+		break;
+	case AT_SET_TRANSPORT_MODE:
+		at_sprintf(&desc->cmd, PUI8("d"),
+			   (int32_t)param->transport_mode);
+		break;
+	case AT_SET_CLIENT_TIMEOUT:
+		at_sprintf(&desc->cmd, PUI8("d"),
+			   (int32_t)param->timeout);
+		break;
+	case AT_PING:
+		at_sprintf(&desc->cmd, PUI8("s"), &param->ping_ip);
+		break;
+	default:
+		return;
+	}
+}
+
+/*
+ * Create the AT command with the specified parameters in desc->cmd
+ * The format of the command will be: AT[CMD][OP]<parmas>\r\n
+ */
+static void build_cmd(struct at_desc *desc, enum at_cmd cmd,
+		      enum cmd_operation op, union in_out_param *param)
+{
+	desc->cmd.len = 0;
+	/* AT */
+	at_strcatch(&desc->cmd, 'A');
+	at_strcatch(&desc->cmd, 'T');
+	/* CMD */
+	at_strcat(&desc->cmd, &g_map[cmd].cmd);
+	/* OP */
+	if (op == AT_QUERY_OP)
+		at_strcatch(&desc->cmd, '?');
+	if (op == AT_TEST_OP) {
+		at_strcatch(&desc->cmd, '=');
+		at_strcatch(&desc->cmd, '?');
+	}
+	if (op == AT_SET_OP && param) {
+		at_strcatch(&desc->cmd, '=');
+		concat_cmd_param(desc, cmd, &param->in);
+	}
+	/* \r\n */
+	at_strcatch(&desc->cmd, '\r');
+	at_strcatch(&desc->cmd, '\n');
+}
+
+/* Send ATE0 command to stop echo */
+static int32_t stop_echo(struct at_desc *desc)
+{
+	uart_write(desc->uart_desc, (uint8_t *)"ATE0\r\n", 6);
+
+	if (SUCCESS != wait_for_response(desc))
+		return FAILURE;
+	desc->result.len = 0;
+
+	return SUCCESS;
+}
+
+/* Handle special cases */
+static int32_t handle_special(struct at_desc *desc, enum at_cmd cmd)
+{
+	switch (cmd) {
+	case AT_RESET:
+		desc->callback_operation = RESETTING_MODULE;
+		uart_write(desc->uart_desc, desc->cmd.buff, desc->cmd.len);
+		mdelay(RESET_DELAY); //Maybe is better wait for ready
```

Comment:
Can you wait using the callback? Or is the reset always under 6 seconds by specification?

Reply:
I can wait for "ready" msg. 6 seconds was a hardcoded value

Reply:
done